File: 2022/python/python/2.py
def part_1(rounds):
    strategy = Strategy(rounds)
    strategy.decrypt(key="moves")
    strategy.calculate_score()
    return strategy.score


def part_2(rounds):
    strategy = Strategy(rounds)
    strategy.decrypt(key="outcomes")
    strategy.calculate_score()
    return strategy.score


def main():
    with open("../../data/2.txt", "r") as f:
        rounds = [[line[0], line[-2]] for line in f.readlines()]

    part_1_answer = part_1(rounds)
    print(f"Part 1: {part_1_answer}")

    part_2_answer = part_2(rounds)
    print(f"Part 2: {part_2_answer}")


class Move:
    def __eq__(self, other):
        if isinstance(other, type(self)):
            return True


class Rock(Move):
    points = 1

    def __lt__(self, other):
        if type(other) == Paper:
            return True
        else:
            return False

    def __gt__(self, other):
        if type(other) == Scissors:
            return True
        else:
            return False


class Paper(Move):
    points = 2

    def __lt__(self, other):
        if type(other) == Scissors:
            return True
        elif type(other) == Rock:
            return False

    def __gt__(self, other):
        if type(other) == Rock:
            return True
        else:
            return False


class Scissors(Move):
    points = 3

    def __lt__(self, other):
        if type(other) == Rock:
            return True
        elif type(other) == Paper:
            return False

    def __gt__(self, other):
        if type(other) == Paper:
            return True
        else:
            return False


class Strategy:
    def __init__(self, rounds: list[list]):
        self.rounds = rounds
        self.moves = []
        self.score = 0

    def decrypt(self, key):
        for match in self.rounds:
            if match[0] == "A":
                opponent_move = Rock()
            elif match[0] == "B":
                opponent_move = Paper()
            elif match[0] == "C":
                opponent_move = Scissors()

            if match[1] == "X":
                if key == "moves":
                    counter_move = Rock()
                elif key == "outcomes":
                    for selected in [Rock, Paper, Scissors]:
                        if opponent_move > selected():
                            counter_move = selected()

            elif match[1] == "Y":
                if key == "moves":
                    counter_move = Paper()
                elif key == "outcomes":
                    for selected in [Rock, Paper, Scissors]:
                        if opponent_move == selected():
                            counter_move = selected()

            else:
                if key == "moves":
                    counter_move = Scissors()
                elif key == "outcomes":
                    for selected in [Rock, Paper, Scissors]:
                        if opponent_move < selected():
                            counter_move = selected()

            self.moves.append([opponent_move, counter_move])

    def calculate_score(self):
        for move in self.moves:
            if move[0] < move[1]:
                self.score += move[1].points + 6
            elif move[0] == move[1]:
                self.score += move[1].points + 3
            elif move[0] > move[1]:
                self.score += move[1].points


if __name__ == "__main__":
    main()

File: 2022/rust/src/bin/2.rs
use aoc::PuzzleInput;

fn main() {
    let p = PuzzleInput::new("../data/2.txt");
    let lines = p.lines();

    let part_1 = part_1(&lines);
    println!("Part 1: {part_1}");

    let part_2 = part_2(&lines);
    println!("Part 2: {part_2}")
}

fn part_1(lines: &Vec<&str>) -> u32 {
    let mut score = 0;
    for line in lines {
        let split: Vec<&str> = line.split_whitespace().collect();

        let opponent = match split[0] {
            "A" => Move::Rock,
            "B" => Move::Paper,
            "C" => Move::Scissors,
            _ => panic!("Invalid move for opponent"),
        };

        let player = match split[1] {
            "X" => Move::Rock,
            "Y" => Move::Paper,
            "Z" => Move::Scissors,
            _ => panic!("Invalid move for player"),
        };

        let outcome = match player {
            _ if player.beats() == opponent => Outcome::Win,
            _ if player.beaten_by() == opponent => Outcome::Lose,
            _ => Outcome::Draw,
        };

        score += outcome.score() + player.score();
    }
    score
}

fn part_2(lines: &Vec<&str>) -> u32 {
    let mut score: u32 = 0;
    for line in lines {
        let split: Vec<&str> = line.split_whitespace().collect();

        let opponent_move = match split[0] {
            "A" => Move::Rock,
            "B" => Move::Paper,
            "C" => Move::Scissors,
            _ => panic!("Invalid move"),
        };

        let outcome = match split[1] {
            "X" => Outcome::Lose,
            "Y" => Outcome::Draw,
            "Z" => Outcome::Win,
            _ => panic!("Invalid outcome"),
        };

        let player_move = match outcome {
            Outcome::Win => opponent_move.beaten_by(),
            Outcome::Lose => opponent_move.beats(),
            Outcome::Draw => opponent_move,
        };

        score += player_move.score() + outcome.score();
    }
    score
}

enum Outcome {
    Lose,
    Draw,
    Win,
}

impl Outcome {
    const fn score(&self) -> u32 {
        match self {
            Self::Lose => 0,
            Self::Draw => 3,
            Self::Win => 6,
        }
    }
}

#[derive(PartialEq, Eq)]
enum Move {
    Rock,
    Paper,
    Scissors,
}

impl Move {
    const fn beaten_by(&self) -> Self {
        match self {
            Self::Rock => Self::Paper,
            Self::Paper => Self::Scissors,
            Self::Scissors => Self::Rock,
        }
    }

    const fn beats(&self) -> Self {
        match self {
            Self::Rock => Self::Scissors,
            Self::Paper => Self::Rock,
            Self::Scissors => Self::Paper,
        }
    }

    const fn score(&self) -> u32 {
        match self {
            Self::Rock => 1,
            Self::Paper => 2,
            Self::Scissors => 3,
        }
    }
}
